solidity/contracts/converter/BancorConverter.sol

```diff
@@ -789,6 +782,174 @@ contract BancorConverter is IBancorConverter, TokenHandler, SmartTokenController
             safeTransferFrom(_token, _from, _to, _amount);
     }
 
+    function addLiquidity(IERC20Token[] memory _reserveTokens, uint256[] memory _reserveAmounts, uint256 _supplyMinReturnAmount)
+        public
+        payable
+        multipleReservesOnly
+    {
+        verifyLiquidityInput(_reserveTokens, _reserveAmounts, _supplyMinReturnAmount);
+
+        for (uint256 i = 0; i < _reserveTokens.length; i++)
+            if (_reserveTokens[i] == address(0))
+                require(_reserveAmounts[i] == msg.value);
+
+        if (msg.value > 0)

```

Comment:
probably better to check this before the previous for loop to save some gas on reverts

Reply:
Up until here, it's all a bunch of `require` statements.
Does it really make any difference which one we do first?
I mean, moving this to be the first means that it will save some gas on some cases of erroneous input, and consume more gas on other cases of erroneous input.
As long as we do them all at the beginning, I don't think it matters which order we do them at (also, this is really meant for very basic errors, I don't think that we should work too hard other than making sure nothing gets executed).

Reply:
the contract factory creation is 4M gas - that's what I think we should do after all verifications

Reply:
@yudilevi : Sorry, but what does that have to do with `addLiquidity`?

Reply:
ah sorry :) there's some storage var reads there, that's 800 gas each, that's the reason I mentioned it here

Reply:
@yudilevi : please see updated code.